### PR TITLE
DM-44095: Implement queryDimensionRecords for RemoteButler

### PR DESCRIPTION
**Pull request by dhirving**
Implemented a large subset of `registry.queryDimensionsRecords` for `RemoteButler` by forwarding to the new query system.

## Checklist

- [ ] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

### CODE REVIEW

File: python/lsst/daf/butler/queries/_expression_strings.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+        Predicate corresponding to that filter, for use in `QueryTree`.
+    """
+    if bind is None:
+        bind = {}
+
+    try:
+        tree = parse_expression(expression)
+    except Exception as exc:
+        raise InvalidQueryError(f"Failed to parse expression '{expression}'") from exc
+
```
Comment (dhirving) [inline python/lsst/daf/butler/queries/_expression_strings.py:81]: The original implementation does some additional processing on the tree here.  I'm not sure if that is stuff already handled by the new query system's validation logic or if we need to handle it here too.

```
    expr = NormalFormExpression.fromTree(tree, NormalForm.DISJUNCTIVE)
    # Check the expression for consistency and completeness.
    visitor = CheckVisitor(data_id, dimensions, bind, defaults, allow_orphans=allow_orphans)
```

File: python/lsst/daf/butler/queries/_expression_strings.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+class _Sequence(NamedTuple):
+    value: list[ColumnExpression]
+    column_type: Literal["sequence"] = "sequence"
+
+
+_VisitorResult: TypeAlias = Predicate | _ColExpr | _Null | _RangeLiteral | _Sequence
+
+
+class _ConversionVisitor(TreeVisitor[_VisitorResult]):
```
Comment (dhirving) [inline python/lsst/daf/butler/queries/_expression_strings.py:129]: The vast majority of the coverage for the query expression language is on `queryDataIds`, so I'm going to implement that next instead of trying to replicate the tests using `queryDimensionRecords`.

The main unimplemented features here are `IN` and miscellaneous `Time`/`Timespan` conversions in `visitBinaryOp`.

There is missing test coverage in the original implementation as well, so I will backfill the tests and complete this implementation in a follow-up to the implementation of queryDataIds.

File: python/lsst/daf/butler/registry/tests/_registry.py [hunk 3]
```diff
@@ -3212,18 +3236,19 @@ def do_query(element, datasets=None, collections=None):
...
             Test("detector", "raft,-name_in_raft", (2, 1, 4, 3)),
             Test("detector", "-detector.purpose", (4,), limit=(1,)),
-            Test("detector", "-purpose,detector.raft,name_in_raft", (2, 3), limit=(2, 2)),
             Test("visit", "visit", (1, 2)),
             Test("visit", "-visit.id", (2, 1)),
             Test("visit", "zenith_angle", (1, 2)),
             Test("visit", "-visit.name", (2, 1)),
-            Test("visit", "day_obs,-timespan.begin", (2, 1)),
-        )
+            Test("visit", "day_obs,-visit.timespan.begin", (2, 1)),
```
Comment (dhirving) [inline python/lsst/daf/butler/registry/tests/_registry.py:3248]: When day_obs became its own dimension, this query became ambiguous because `timespan` is present in both day_obs and visit (with different meaning/values).  The new query system correctly threw an exception complaining about the ambiguity, but the old one silently picks the visit one.

File: python/lsst/daf/butler/registry/tests/_registry.py [hunk 4]
```diff
@@ -3386,7 +3423,7 @@ def testExposureQueries(self):
             [
                 record.id
                 for record in registry.queryDimensionRecords("visit", instrument="HSC")
-                .order_by("id")
+                .order_by("visit")
```
Comment (dhirving) [inline python/lsst/daf/butler/registry/tests/_registry.py:3434]: Similar to above, `id` became ambiguous when `day_obs` became a dimension.

File: python/lsst/daf/butler/registry/tests/_registry.py [hunk 5]
```diff
@@ -106,6 +107,22 @@ class RegistryTests(ABC):
     """True if the registry class being tested supports regex searches for
     collections."""
 
+    supportsDetailedQueryExplain: bool = True
```
Comment (dhirving) [inline python/lsst/daf/butler/registry/tests/_registry.py:110]: This is mostly stuff that requires `LIMIT 1` execution of subqueries to diagnose.  This is marked as a TODO in DirectQueryDriver, so we can flip this feature flag to get these tests when someone goes to backfill that feature.

File: python/lsst/daf/butler/registry/tests/_registry.py [hunk 6]
```diff
@@ -106,6 +107,22 @@ class RegistryTests(ABC):
...
+    explanations for queries that return no rows by running additional queries
+    to diagnose the problem.
+    """
+
+    supportsQueryOffset: bool = True
+    """True if the registry class being tested supports the 'offset' parameter
+    to query methods.
+    """
+
+    supportsQueryGovernorValidation: bool = True
```
Comment (dhirving) [inline python/lsst/daf/butler/registry/tests/_registry.py:121]: I think this just wasn't implemented for the new query system?  Not sure how important/desirable this validation is -- I think it's more consistent to just get no results here instead of sometimes throwing an exception depending which field you tried to filter on.

File: python/lsst/daf/butler/queries/_expression_strings.py (hunk 1 above)
Comment (TallJimbo) [reply python/lsst/daf/butler/queries/_expression_strings.py:81]: This stuff should indeed already be handled.

File: python/lsst/daf/butler/registry/tests/_registry.py (hunk 6 above)
Comment (TallJimbo) [reply python/lsst/daf/butler/registry/tests/_registry.py:121]: We've gone back and forth on whether to raise or get no results here, due to conflicting requests from users (which we didn't always recognize as conflicting).  For the new query system I came down on the side of just returning no results, both because I also think that's more consistent and because in the new system (especially thinking about `RemoteButler`) it looked harder to have the exception raised at the point at which an error message would be most useful.

I've added this to the list of changes to be RFC'd on [DM-43587](https://rubinobs.atlassian.net/browse/DM-43587).

File: python/lsst/daf/butler/queries/_expression_strings.py (hunk 2 above)
Comment (TallJimbo) [reply python/lsst/daf/butler/queries/_expression_strings.py:129]: Did `IN` and the `Time`/`Timespan` operators look like they were impossible to represent in the new `Predicate` tree, or did you just not want to do them yet?

Another way to approach the testing is that the coverage for `Predicate` trees and their mapping to SQL should now be pretty decent via the new query system's tests, so just testing the conversion from string to `Predicate` tree is probably sufficient for poking at all of the edge cases.

File: python/lsst/daf/butler/queries/_expression_strings.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+                # I don't know how to provide.
+                raise NotImplementedError("INGEST_DATE not supported yet")
+            case ExpressionConstant.NULL:
+                return _Null()
+            case None:
+                pass
+            case _:
+                raise AssertionError("Unhandled expression constant")
+
+        element, column = categorizeElementId(self.universe, name)
```
Comment (TallJimbo) [inline python/lsst/daf/butler/queries/_expression_strings.py]: I'd prefer to delegate to `lsst.daf.butler.queries.convert_args.interpret_identifier` here.  That will require initializing the visitor with more information about the state of the query the expression is being added to, but that'll solve the problem of how to handle `INGEST_DATE`, provide hopefully-better error messages, and make `where` and `order_by` identifier intepretation more consistent.

(I actually think it will only recognize lower-case `ingest_date`, so we may need some case-wrangling in `interpret_identifier).

File: python/lsst/daf/butler/queries/_expression_strings.py [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+        except ValueError:
+            # int() raises for float-like strings
+            numeric = float(value)
+        return _make_literal(numeric)
+
+    def visitParens(self, expression: _VisitorResult, node: Node) -> _VisitorResult:
+        return expression
+
+    def visitPointNode(self, ra: _VisitorResult, dec: _VisitorResult, node: Node) -> _VisitorResult:
+        raise NotImplementedError("POINT() function is not supported yet")
```
Comment (TallJimbo) [inline python/lsst/daf/butler/queries/_expression_strings.py:211]: I'm pretty confident it's never actually been supported, at least not beyond the parser.

File: python/lsst/daf/butler/remote_butler/registry/_query_dimension_records.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    @property
+    def element(self) -> DimensionElement:
+        return self._args.element
+
+    def __iter__(self) -> Iterator[DimensionRecord]:
+        with self._build_query() as result:
+            # We have to eagerly fetch the results to prevent
+            # leaking the resources associated with QueryDriver.
+            records = list(result)
+        yield from records
```
Comment (TallJimbo) [inline python/lsst/daf/butler/remote_butler/registry/_query_dimension_records.py]: I suspect (but do not know) that `return iter(records)` is a little more efficient in cases like this.

File: python/lsst/daf/butler/remote_butler/registry/_query_dimension_records.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        with self._query_factory() as query:
+            a = self._args
+            for dataset_type in a.dataset_types:
+                query = query.join_dataset_search(dataset_type, a.collections)
+            if a.where:
+                query = query.where(a.where, bind=a.bind)
+            if a.dataId:
+                query = query.where(a.dataId)
+            if a.kwargs:
+                query = query.where(a.kwargs)
```
Comment (TallJimbo) [inline python/lsst/daf/butler/remote_butler/registry/_query_dimension_records.py]: We should combine these last two cases into one:
```
if a.dataId:
     kwargs = a.kwargs if a.kwargs else {}  # or use default_factory=dict in dataclass
     query = query.where(a.dataId, **a.kwargs)
```
This lets the values in `kwargs` replace those in `dataId`, which is what users have come to expect from data IDs in kwargs, instead of putting the conflicting constraint into the query and dooming it yield no results.

**Review by TallJimbo — APPROVED** [review]

File: python/lsst/daf/butler/queries/_expression_strings.py (hunk 2 above)
Comment (dhirving) [reply python/lsst/daf/butler/queries/_expression_strings.py:129]: At first glance they appear to be relatively straightforward to implement -- there are just a lot of cases and no test coverage, so I wanted to defer to a later PR with less going on.
